File: packages/contracts/src/canonical.ts
export type JsonPrimitive = string | number | boolean | null;
export type JsonValue = JsonPrimitive | JsonValue[] | { [key: string]: JsonValue };

const HEX_CONTROL = "0123456789abcdef";

function escapeJsonString(value: string): string {
  let out = "\"";
  for (const char of value) {
    const code = char.codePointAt(0);
    if (code === undefined) {
      continue;
    }
    if (char === "\\") {
      out += "\\\\";
    } else if (char === "\"") {
      out += "\\\"";
    } else if (code < 0x20) {
      const hex = code.toString(16).padStart(4, "0");
      if ([...hex].some((h) => !HEX_CONTROL.includes(h))) {
        throw new Error("canonical json: invalid control hex");
      }
      out += `\\u${hex}`;
    } else {
      out += char;
    }
  }
  out += "\"";
  return out;
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

export function canonicalJson(value: unknown): string {
  if (value === null) {
    return "null";
  }
  if (value === true) {
    return "true";
  }
  if (value === false) {
    return "false";
  }
  if (typeof value === "number") {
    if (!Number.isSafeInteger(value)) {
      throw new Error("canonical json: only safe integers are allowed");
    }
    return String(value);
  }
  if (typeof value === "string") {
    return escapeJsonString(value);
  }
  if (Array.isArray(value)) {
    return `[${value.map((item) => canonicalJson(item)).join(",")}]`;
  }
  if (isPlainObject(value)) {
    const keys = Object.keys(value).sort();
    const fields = keys.map((key) => `${escapeJsonString(key)}:${canonicalJson(value[key])}`);
    return `{${fields.join(",")}}`;
  }
  throw new Error("canonical json: unsupported value");
}

File: src/evidenceops/contracts/canonical.py
from __future__ import annotations

from typing import Any

_HEX = "0123456789abcdef"


def _escape_json_string(value: str) -> str:
    out: list[str] = ['"']
    for char in value:
        code = ord(char)
        if char == "\\":
            out.append("\\\\")
        elif char == '"':
            out.append('\\"')
        elif code < 0x20:
            hex_digits = f"{code:04x}"
            if any(digit not in _HEX for digit in hex_digits):
                raise ValueError("canonical json: invalid control hex")
            out.append(f"\\u{hex_digits}")
        else:
            out.append(char)
    out.append('"')
    return "".join(out)


def canonical_json(value: Any) -> str:
    if value is None:
        return "null"
    if value is True:
        return "true"
    if value is False:
        return "false"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, int) and not isinstance(value, bool):
        if abs(value) > 9007199254740991:
            raise ValueError("canonical json: only safe integers are allowed")
        return str(value)
    if isinstance(value, str):
        return _escape_json_string(value)
    if isinstance(value, list):
        return "[" + ",".join(canonical_json(item) for item in value) + "]"
    if isinstance(value, dict):
        keys = sorted(value.keys())
        fields = [_escape_json_string(key) + ":" + canonical_json(value[key]) for key in keys]
        return "{" + ",".join(fields) + "}"
    raise ValueError("canonical json: unsupported value")
